src/librustc_mir/borrow_check/nll/region_infer/mod.rs

```diff
@@ -1128,27 +1195,57 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         let constraints = self.find_constraint_paths_from_region(fr.clone());
         let path = constraints.iter().min_by_key(|p| p.len()).unwrap();
         debug!("report_error: path={:?}", path);
+
         let path = path.iter()
             .filter(|index| self.constraint_is_interesting(index))
             .collect::<Vec<&ConstraintIndex>>();
         debug!("report_error: path={:?}", path);
 
-        let fr_string = match fr_name {
-            Some(r) => format!("free region `{}`", r),
-            None => format!("free region `{:?}`", fr),
-        };
+        let mut categorized_path = path.iter().filter_map(|index| {
+            self.constraints.get(**index).iter().filter_map(|constraint| {
+                let span = constraint.locations.span(mir);
+                constraint.locations.from_location().iter().filter_map(|location| {
+                    let classification = self.classify_constraint(*location, mir);
+                    Some((classification, span))
+                }).next()
+            }).next()
+        }).collect::<Vec<(ConstraintCategory, Span)>>();
+        debug!("report_error: categorized_path={:?}", categorized_path);
+
+        categorized_path.sort_by(|p0, p1| p0.0.cmp(&p1.0));
+        debug!("report_error: sorted_path={:?}", categorized_path);
+
+        if categorized_path.len() > 0 {
+            let blame_constraint = &categorized_path[0];
+
+            let mut diag = infcx.tcx.sess.struct_span_err(
```

Comment:
can I see the output of this code for any of our tests?

Reply:
For the test we focused on in the pair programming session, this is the current output:
```
error: Cast
  --> src/test/ui/underscore-lifetime/dyn-trait-underscore.rs:18:5
   |
16 |   fn a<T>(items: &[T]) -> Box<dyn Iterator<Item=&T>> {
   |  ____________________________________________________-
17 | |     //                      ^^^^^^^^^^^^^^^^^^^^^ bound *here* defaults to `'static`
18 | |     Box::new(items.iter()) //~ ERROR cannot infer an appropriate lifetime
   | |     ^^^^^^^^^^^^^^^^^^^^^^
   | |     |
   | |     Cast
   | |     Cast
19 | | }
   | |_- Cast
```
I don't know if it's great, because we end up with a lot of `Cast`.